### ISSUE TITLE
CPU Usage triples after upgrade to v2.1.0

### ISSUE THREAD

**Issue by dannyk81** [prometheus/prometheus#3715]
**What did you do?**

Upgraded Prometheus from v2.0.0 to v2.1.0 (using the official Docker image)

**What did you expect to see?**

Similar or better performance figures.

**What did you see instead? Under which circumstances?**

CPU Usage almost tripled and rule evaluation times increased, we observed the same on two different deployment (two separate K8s clusters), after failing back to v2.0.0 figures are back to normal.

Cluster 1 [CPU]:
![image](https://user-images.githubusercontent.com/24274156/35181757-cf981d80-fd95-11e7-8409-5cda26f877c9.png)

Cluster 2 [CPU]:
![image](https://user-images.githubusercontent.com/24274156/35181776-2aff81cc-fd96-11e7-81f0-05de70532a5c.png)

Cluster 1 [Rule evluation duration]:
![image](https://user-images.githubusercontent.com/24274156/35181828-431f6334-fd97-11e7-9b2a-58e679e03535.png)

Cluster 2 [Rule evluation duration]: graph looks about the same, but this cluster is smaller:
![image](https://user-images.githubusercontent.com/24274156/35181839-84c86984-fd97-11e7-82a4-7e78859b82a9.png)

Except changing the docker image version, everything remained unchanged.

**Environment**

* System information:

`Linux 4.12.10-coreos x86_64` (Runs in K8s 1.5.3 Cluster)

* Prometheus version:

```
prometheus, version 2.1.0 (branch: HEAD, revision: 85f23d82a045d103ea7f3c89a91fba4a93e6367a)
  build user:       root@6e784304d3ff
  build date:       20180119-12:01:23
  go version:       go1.9.2
```

* Alertmanager version:

	N/A

* Prometheus configuration file:
```
Can provide if needed
```

* Alertmanager configuration file:
```
N/A
```

* Logs:
```
Can provide if needed
```

**Comment by JorritSalverda** [comment]
I see the same thing on Kubernetes Engine 1.8.6-gke.0 going from Prometheus 2.0.0 to 2.1.0:

![20180120-prometheus-210-cpu-increase](https://user-images.githubusercontent.com/136472/35184120-a1d6b6f2-fdf1-11e7-81f3-33a9ad23cff7.png)

The logs don't indicate there's anything special going on:

```
level=info ts=2018-01-20T13:09:22.585447311Z caller=main.go:225 msg="Starting Prometheus" version="(version=2.1.0, branch=HEAD, revision=85f23d82a045d103ea7f3c89a91fba4a93e6367a)"
level=info ts=2018-01-20T13:09:22.585523718Z caller=main.go:226 build_context="(go=go1.9.2, user=root@6e784304d3ff, date=20180119-12:01:23)"
level=info ts=2018-01-20T13:09:22.585546075Z caller=main.go:227 host_details="(Linux 4.4.86+ #1 SMP Thu Dec 7 20:11:11 PST 2017 x86_64 prometheus-1 (none))"
level=info ts=2018-01-20T13:09:22.585564529Z caller=main.go:228 fd_limits="(soft=1048576, hard=1048576)"
level=info ts=2018-01-20T13:09:22.588658792Z caller=main.go:499 msg="Starting TSDB ..."
level=info ts=2018-01-20T13:09:22.588686556Z caller=web.go:383 component=web msg="Start listening for connections" address=0.0.0.0:9090
level=info ts=2018-01-20T13:09:50.424774629Z caller=main.go:509 msg="TSDB started"
level=info ts=2018-01-20T13:09:50.42494659Z caller=main.go:585 msg="Loading configuration file" filename=/prometheus-config/prometheus.yml
level=info ts=2018-01-20T13:09:50.432711846Z caller=kubernetes.go:191 component="discovery manager scrape" discovery=k8s msg="Using pod service account via in-cluster config"
level=info ts=2018-01-20T13:09:50.433969004Z caller=kubernetes.go:191 component="discovery manager scrape" discovery=k8s msg="Using pod service account via in-cluster config"
level=info ts=2018-01-20T13:09:50.434900141Z caller=kubernetes.go:191 component="discovery manager scrape" discovery=k8s msg="Using pod service account via in-cluster config"
level=info ts=2018-01-20T13:09:50.435884965Z caller=kubernetes.go:191 component="discovery manager scrape" discovery=k8s msg="Using pod service account via in-cluster config"
level=info ts=2018-01-20T13:09:50.437116695Z caller=kubernetes.go:191 component="discovery manager scrape" discovery=k8s msg="Using pod service account via in-cluster config"
level=info ts=2018-01-20T13:09:50.438063243Z caller=kubernetes.go:191 component="discovery manager scrape" discovery=k8s msg="Using pod service account via in-cluster config"
level=info ts=2018-01-20T13:09:50.438998843Z caller=kubernetes.go:191 component="discovery manager scrape" discovery=k8s msg="Using pod service account via in-cluster config"
level=info ts=2018-01-20T13:09:50.459854142Z caller=main.go:486 msg="Server is ready to receive web requests."
level=info ts=2018-01-20T13:09:50.459854122Z caller=manager.go:59 component="scrape manager" msg="Starting scrape manager..."
```

**Comment by FUSAKLA** [comment]
I see the same thing in our kubernetes cluster running official docker image and also changing only the image tag. 
(the firts annotation is upgrade from `2.0.0` to `2.1.0` the second one is restart due to the OOM kill by kuberrnetes)

![allocations](https://user-images.githubusercontent.com/6112562/35192823-8a9cc140-fe99-11e7-8750-cc3c703eb433.png) 
![compaction](https://user-images.githubusercontent.com/6112562/35192825-93f7fa66-fe99-11e7-922b-b84b81c2da3a.png)
![cpu](https://user-images.githubusercontent.com/6112562/35192826-96a48a40-fe99-11e7-92b4-9dfef3d23069.png)
![federation](https://user-images.githubusercontent.com/6112562/35192827-9945388a-fe99-11e7-802e-d0b41b0040be.png)
This is interesting because according to the release note the federation should be faster
```
[ENHANCEMENT] Federation is now much faster for large numbers of series.
```
To provide HA I federate all timeseries (I know it's not right it's just temporary solution) so it is I think exactly this case as stated and the federation is even slower.

![memory](https://user-images.githubusercontent.com/6112562/35192829-9bbca7f6-fe99-11e7-92c8-7069e8d6ad4b.png)
![reloaded-blocks](https://user-images.githubusercontent.com/6112562/35192831-9eb12130-fe99-11e7-83d6-0bdaf40c9a98.png)
![wal-fsync](https://user-images.githubusercontent.com/6112562/35192832-a0e42f1a-fe99-11e7-921b-550813e0f500.png)

Recording rules evaluation time did not change in my case.

**Comment by Gouthamve** [comment]
Running a test setup where I am federating 380K series, there is indeed a CPU gain and allocs increase of about 20% and slight slowdown (~10%). While nowhere close to the numbers reported, this certainly mandates a look. 

This is one of the changes in the federation code-path: https://github.com/prometheus/prometheus/pull/3569

**Comment by goddog1118** [comment]
may i ask what's the min time interval of this tool ? can i set scrape_interval to sub second?

**Comment by brian-brazil** [comment]
@goddog1118 Please don't ask unrelated questions on bugs, use the [prometheus-users mailing list](https://groups.google.com/forum/#!forum/prometheus-users).

**Comment by paddynewman** [comment]
I'm also seeing similar problems, in case it helps with prioritisation.

Here's a graph showing the difference in resource utilisation at our site. The only change was to upgrade from 2.0 to 2.1:

![prometheus 2 0 to 2 1](https://user-images.githubusercontent.com/101397/35326846-d74a51e4-00ef-11e8-8228-ff5994215e6a.png)

**Comment by brian-brazil** [comment]
Can someone share their configuration and a pprof?

**Comment by paddynewman** [comment]
Here's our configuration. I've had to heavily redact it; I hope it's still useful. The codeshare will expire in 24 hours.

https://codeshare.io/2B8zmd

**Comment by krasi-georgiev** [comment]
@paddynewman thanks I will try to look into this in the next few days and will let you know if I find anything.
mem and cpu pprof would really help narrow it down.
If it is not too much to ask please generate mem and cpu pprof with 2.0 (normal performance) and 2.1 higher cpu and mem usage to compare the results.

**Comment by brancz** [comment]
@FUSAKLA the graph you showed on memory allocations also looks interesting.

In general as @krasi-georgiev mentioned cpu profile as well as memory allocation profiles would be great. If you have a go runtime installed you can collect a memory profile:

```
go tool pprof http://<your prometheus server>/debug/pprof/heap
```

And a 30s CPU profile with

```
go tool pprof http://<your prometheus server>/debug/pprof/profile
```

These commands will write profiles to disk, you can upload that profile for us to analyze.

**Comment by paddynewman** [comment]
I have the profiling output, at least for 2.1, but my $employer will not allow me to share it. I can generate the output for 2.0 as well but again, I won't be allowed to share it.

Is there anything else I can do to help while I figure out how to use pprof myself?

**Comment by brian-brazil** [comment]
Profiling output will not contain anything sensitive, it's mainly function names.

**Comment by JorritSalverda** [comment]
I can create those profiles and share them but have no experience using pprof. When I run it it puts me in an interactive mode. Are there any flags to pass so it collects what you need in a non-interactive way?

**Comment by brancz** [comment]
you can leave the interactive mode right away, the first line or so should say that the full profile has been written to `~/pprof/<profile-name>`.

**Comment by paddynewman** [comment]
Prometheus 2.1 profiles:

https://drive.google.com/folderview?id=13_Scu9wS2u_odOyXvbhJMAZsgXk9yOIr

**Comment by JorritSalverda** [comment]
I've created profiles for both versions at although from different servers. None of the servers uses federation.

https://drive.google.com/file/d/1fXE64oJ8uA86a47nqD_23BQ2nuaKB6XR/view?usp=sharing

**Comment by FUSAKLA** [comment]
Sorry for the delay

I managed to get pprof data from the federated `2.1.0` prometheus instance 
Unfortunately the instance got OOM kill recently so the memory usage is not SO high. 
I'll get more critical data tomorrow if needed.
[link here](https://drive.google.com/open?id=1eZQRs2_DP1eDtiYiHE6JVh-CGwzAkRiA) (I hope i don't break any rules in my company :) )

i also added prometheus metrics and configurations (both the federated and the federating instances)

Sadly a tried to downgrade back to `2.0.0` version but end up with 
```
level=info ts=2018-01-24T17:32:57.106618248Z caller=main.go:314 msg="Starting TSDB"
level=info ts=2018-01-24T17:32:57.106646142Z caller=targetmanager.go:71 component="target manager" msg="Starting target manager..."
level=error ts=2018-01-24T17:32:57.187507378Z caller=main.go:323 msg="Opening storage failed" err="read meta information /prometheus/01C4MHVPYRN7THY4KS3KAZP1A5: unexpected meta file version 2"
```

the metafile.json contains:
```
{
	"ulid": "01C4MHVPYRN7THY4KS3KAZP1A5",
	"minTime": 1516802400000,
	"maxTime": 1516809600000,
	"stats": {
		"numSamples": 13784198,
		"numSeries": 42773,
		"numChunks": 117088
	},
	"compaction": {
		"level": 1,
		"sources": [
			"01C4MHVPYRN7THY4KS3KAZP1A5"
		]
	},
	"version": 2
}
```

So it looks like i have problem with downgrade from `2.1.0` to `2.0.0` but in changelog was no mention about backward incompatibility in storage?

This would mean more problem and minor data loss which won't be so critical but I would prefer to solve this better way. Than I can give the pprof data from `2.0.0` version

**Comment by tomwilkie** [comment]
I suspect this is a side effect of the switch to mergeseriesset over dedupeseriesset - the latter would just drop equivalent series, whilst the former merges them.  This should be easy to test - but can someone share the federation match params they use?  If they have multiple matchers returns the same series that might explain it.

**Comment by tomwilkie** [comment]
If someone experiencing the issue could test the above PR, I think that might fix this.

**Comment by FUSAKLA** [comment]
my federation uses `'{job=~"(.+)"}'` the configuration is in the file i posted

I can try the PR version but i need docker image is it built automatically or do i have to build it by myself?

**Comment by tomwilkie** [comment]
@FUSAKLA thats a single match, so should cause the original issue I suspected, although my PR might speed things up.

I'll build you a docker image.

**Comment by FUSAKLA** [comment]
It's ok i can do it by myself i just didn't know if you have some CI docker repository with branch builds.

**Comment by tomwilkie** [comment]
Just pushed tomwilkie/prometheus:3715-2.1-cpu-749781ed, let me know if it helps!

**Comment by FUSAKLA** [comment]
There is *some* improvement in memory usage but I'm not sure if significant enough to say it caused by the fix. The CPU and allocations seems to by untouched by the fix.
![screenshot from 2018-01-25 12-47-29](https://user-images.githubusercontent.com/6112562/35386893-f2d48ab8-01cd-11e8-8ecb-e6c90c593078.png)

**Comment by tomwilkie** [comment]
Okay, I think that probably rules out the federation changes as the cause - with that fix they should be strictly better than 2.0.  Thanks for your help.

**Comment by FUSAKLA** [comment]
No problem, I'll be glad to help sort this out. I was waiting for those bugfixes introduced in 2.1.0.

**Comment by tinytub** [comment]
https://github.com/prometheus/client_golang/issues/366

maybe the gzip cause the high cpu usage？
federate and client_golang promhttp use same writeSample funcstions.

**Comment by brian-brazil** [comment]
The gzip code hasn't changed in a long time.

**Comment by tinytub** [comment]
sorry I am wrong, 
I take a pprof when `curl "http://mdb73:9090/federate?match[]{__name__=~"disk.*"}" -H 'Accept-Encoding: gzip'`
It's about 637342 line series
https://drive.google.com/file/d/1QLAcmxbbkovc7h9EStlG37_AwvNer4hR/view?usp=sharing

Hope it helps

**Comment by brancz** [comment]
I just opened https://github.com/prometheus/prometheus/pull/3740 which I believe fixes the primary issue (and elaborates on another one, which shouldn't be a problem though except for higher allocations). If someone could test it, that would be great! :slightly_smiling_face:

**Comment by FUSAKLA** [comment]
Actually SD went pretty crazy also after the update but I didnt consider it so important.
There is graph of the upgrade influence on SD
![screenshot from 2018-01-25 23-33-08](https://user-images.githubusercontent.com/6112562/35416111-43a9efdc-0228-11e8-9781-d0927729afc1.png)

I'll test the PR asap

**Comment by krasi-georgiev** [comment]
seems that this was the main cause, when I did the refactoring I removed the throttling which caused the increased mem and cpu usage.

**Comment by FUSAKLA** [comment]
I'm testing it and it looks really promising.
In few minutes I'll post the results. Sorry for not posting the SD figures at the first time.

**Comment by FUSAKLA** [comment]
I think this pretty much confirms it.
All values are back to `2.0.0` normal 
![screenshot from 2018-01-26 00-44-07](https://user-images.githubusercontent.com/6112562/35418385-37636f14-0232-11e8-9e08-82861eb64bec.png)
![screenshot from 2018-01-26 00-44-55](https://user-images.githubusercontent.com/6112562/35418393-4103e4b8-0232-11e8-9c4c-c4a44d5eb793.png)
![screenshot from 2018-01-26 00-44-31](https://user-images.githubusercontent.com/6112562/35418394-41228990-0232-11e8-8261-f162132e5190.png)
even the federation is back to the previous values
![screenshot from 2018-01-26 00-44-21](https://user-images.githubusercontent.com/6112562/35418452-8fad0aea-0232-11e8-91a0-4630a9d0ef7a.png)

Do you want pprof or any other data to verify this?

Thanks for the PR so quickly!

**Comment by paddynewman** [comment]
Do you think this PR will resolve (what assuming is) a memory leak that appeared in 2.1 as well?

![memory-leak](https://user-images.githubusercontent.com/101397/35432930-9b663360-0279-11e8-8925-866a21722ab5.png)

**Comment by FUSAKLA** [comment]
as far as i can tell from metrics probably not.. this figure shows impact on memory since 2.1.0 upgrade and the last annotation is upgrade to the mentioned PR and memory is still rising although the CPU and so on seems to be ok
![screenshot from 2018-01-26 10-23-38](https://user-images.githubusercontent.com/6112562/35433341-672e8d22-0283-11e8-9a17-948fd635e650.png)

**Comment by brancz** [comment]
Are you looking at memory allocated and exposed by the go runtime, or from the view of the operating system? If it's the latter then this looks like a pretty normal GC behavior (albeit still optimizable).

**Comment by FUSAKLA** [comment]
this figure shows `process_resident_memory_bytes` and `go_memstats_alloc_bytes`. And data from cadvisor looks the same.

**Comment by paddynewman** [comment]
I'm fairly certain there's a memory leak. In my previous comment/graph I am showing the OS's view of memory but the consumption never goes down.

I have a midnight restart enabled via cron which is why you see the saw-tooth pattern and this was only required after upgrading from 2.0 to 2.1. Before that the cosumption was very stable at ~ 17.5GiB.

**Comment by brancz** [comment]
After another day of benchmarking we've come up with this: https://github.com/prometheus/prometheus/pull/3747

Please try it out and give us feedback! :slightly_smiling_face:

**Comment by FUSAKLA** [comment]
Thank`s for the effort!

Unfortunately I left the previous PR version deployed so this is not compared to the official `2.1.0` release but to the proof of concept version in https://github.com/prometheus/prometheus/pull/3740

Tomorrow I'll switch to the official `2.1.0` and send those results. 

But again it looks good. Memory usage fell down and hopefully wont spike anymore 
 also reloading blocks from disk mentioned in https://github.com/prometheus/prometheus/issues/3725 fell to minimum
The rest like SD, federation and so on looks the same as in the previous fix
![screenshot from 2018-01-27 00-20-56](https://user-images.githubusercontent.com/6112562/35464760-fbfb03b2-02f7-11e8-9af3-2d7ef49adf20.png)
![screenshot from 2018-01-27 00-19-59](https://user-images.githubusercontent.com/6112562/35464761-fc1b3a42-02f7-11e8-93e0-8c64d033c68c.png)

**Comment by FUSAKLA** [comment]
So i have the results.. There are 3 versions in those figures:

(from left to right separated by annotations)

   1. The first PR already deleted
   2. This PR now requested to benchmark from branch `sched-update-throttle`
   3. official `2.1.0` release

![screenshot from 2018-01-27 15-33-37](https://user-images.githubusercontent.com/6112562/35472933-9d99a9b4-0378-11e8-91fb-178d09568993.png)
![screenshot from 2018-01-27 15-33-28](https://user-images.githubusercontent.com/6112562/35472934-9db7bce2-0378-11e8-8041-25448a58404c.png)
![screenshot from 2018-01-27 15-33-49](https://user-images.githubusercontent.com/6112562/35472932-9d7c3686-0378-11e8-92e6-14213d113673.png)
![screenshot from 2018-01-27 15-34-00](https://user-images.githubusercontent.com/6112562/35472931-9d5dc200-0378-11e8-8670-276e5478ff6f.png)

**Comment by Gouthamve** [comment]
Should be closed in #3747

**Comment by dannyk81** [comment]
Guys! I dropped this bomb on went on vacation 😛  but was amazed to follow remotely how this was tackled and finally resolved. 

One of the main reasons I love Prometheus so much is thanks to the incredible core team and community!!! ❤️ 

Can't wait for the next release with the fixes 👏 👏
